<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Qian Cheng" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Tutorial - G2P</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Tutorial";
        var mkdocs_page_input_path = "Tutorial.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/rust.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> G2P
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../QuickStart/">QuickStart</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Tutorial</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-files-and-data-preprocessing">Input files and data preprocessing</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#format-conversion-of-genotypic-data">Format conversion of genotypic data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#filtering-of-genotypic-data">Filtering of genotypic data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#data-quality-control-of-genomic-selection">Data quality control of genomic selection</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#genomic-selection-and-model-evaluation">Genomic selection and model evaluation</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#genotype-to-phenotype-prediction">Genotype-to-phenotype prediction</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#g2p-cross-validation">G2P cross-validation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#evaluation">Evaluation</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#integration-of-prediction-results">Integration of prediction results</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#two-results-integration-and-user-defined-integration-more-than-two-models-results">Two results integration and user-defined integration (more than two models results)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#multiple-results-integration">Multiple results integration</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#refinement-design-of-training-dataset">Refinement design of training dataset</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../about/">About</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">G2P</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>Tutorial</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/GOVS-env/G2P/edit/master/docs/Tutorial.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="tutorial"><center>Tutorial</center></h1>
<p><center>Introduction of main functions and parameter information</center></p>
<hr />
<p>In this section, the main functions in G2P and thier parameters will be introducted in details.</p>
<h2 id="input-files-and-data-preprocessing">Input files and data preprocessing</h2>
<h3 id="format-conversion-of-genotypic-data">Format conversion of genotypic data</h3>
<ul>
<li><em><strong><code>GSTransForm</code></strong></em></li>
<li>Description</br>
The function can only be used in G2P-sif.By calling other external software like VCFtools Plink TASSEL etc for the transition of genotype format</li>
<li>Usage <br/>
<code>GSTransForm(pattern, input, output, workdir = "./", memory = 8 )</code></li>
<li>Arguments<ul>
<li><code>pattern</code>(character) pattern of transformation, including "hmp2plink","plink2vcf","vcf2hmp","vcf2plink","plink2hmp", and "hmp2vcf".</li>
<li><code>input</code>(character) name of original file, prefix for plink, full name for hapmap and vcf.</li>
<li><code>output</code>(character) name of original file, prefix for all.</li>
<li><code>workdir</code>(character) set working directory</li>
<li><code>memory</code>(integer) set memory size 
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">load(cubic_hmp)
setwd(&quot;./&quot;)
write.table(hmp,file = &quot;test.hmp.txt&quot;,row.names = F,col.names = T,sep = &quot;\t&quot;,quote = F)

## transformation ##
GSTransForm(pattern=&quot;hmp2plink&quot;,input=&quot;test.hmp.txt&quot;,output=&quot;h2p&quot;,workdir=&quot;./&quot;)
</code></pre>
<h3 id="filtering-of-genotypic-data">Filtering of genotypic data</h3>
<ul>
<li><em><strong><code>GSFiltering</code></strong></em></li>
<li>Description<br/>
The function can raduce genotypic data by calling Plink.</li>
<li>Usage <br/>
<code>GSFiltering(input, output , 
            inputformat = "hmp", missingRate = 0, 
            maf = 0.05,window = 100,step = 5,r2 = 0.1)</code></li>
<li>Arguments<ul>
<li><code>input</code>(character) name of original file, prefix for plink, full name for hapmap and vcf.</li>
<li><code>output</code>(character) name of original file, prefix for all.</li>
<li><code>inputformat</code>(character) the in file format of input file, "hmp" for hapmap,"vcf" VCF or "plink" for bed,bim,fam.</li>
<li><code>missingRate</code>(numeric [0-1]) filters out all variants with missing call rates exceeding the provided value to be removed</li>
<li><code>maf</code>(numeric [0-1]) filters out all variants with minor allele frequency below the provided threshold (default 0.05)</li>
<li><code>window</code>(numeric or character) window size in variant count (numeric) or kilobase (if the 'kb' modifier is present). </li>
<li><code>step</code>(numeric) variant count to shift the window at the end of each step</li>
<li><code>r2</code>(numeric) pairwise r2 threshold: at each step.</li>
<li><code>workdir</code>(character) set working directory</li>
<li><code>memory</code>(integer) set memory size 
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">## load example data ##
load(cubic_hmp)
setwd(&quot;./&quot;)
write.table(hmp,file = &quot;test.hmp.txt&quot;,row.names = F,col.names = T,sep = &quot;\t&quot;,quote = F)

## filtering 
GSFiltering(input = &quot;test.hmp.txt&quot;, output = &quot;filter_res&quot;, 
            inputformat = &quot;hmp&quot;, missingRate = 0, 
            maf = 0.05,window = 100,step = 5,r2 = 0.1)
</code></pre>
<h3 id="data-quality-control-of-genomic-selection">Data quality control of genomic selection</h3>
<ul>
<li><em><strong><code>GSDataQC</code></strong></em></li>
<li>Description<br/>
The function can raduce genotypic data by calling Plink.</li>
<li>Usage <br/>
<code>GSFiltering(input, output , 
            inputformat = "hmp", missingRate = 0, 
            maf = 0.05,window = 100,step = 5,r2 = 0.1)</code></li>
<li>Arguments<ul>
<li><code>markers</code>(numeric, matrix) genotypic data, row represents sample and column represents SNP (feature). Missing (NA) alleles are allowed</li>
<li><code>phenotype</code>(data.frame) the phenotype data. The first column must be the sample name.</li>
<li><code>pheno</code>Phenotypic data frame, the first column describes sample IDs.</li>
<li><code>impute</code>(logical) if TRUE, imputation, default FALSE.</li>
<li><code>filter</code>(logical) if TRUE, filtering genotypic data with MAF and missing rate.</li>
<li><code>NABound</code>(numeric, [0,1]) max missing value percentage.</li>
<li><code>MAFBound</code>(numeric, [0,1]) min MAF percentage in each marker.</li>
<li><code>imputeMethod</code>(character) the method of imputation, "mean" or "median", default "median".</li>
<li><code>hete</code>(numeric) if genotypic matrix coded with numeric or transformed to numeric from letter, this parameter is used for MAF summary which define the numeric code of heterozygote, default 1 among 0,1,2.</li>
<li><code>silent</code>(logical) TRUE for hiding progress bar and FALSE for showing progress bar.
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">data(MZ)
## generate missing value
misIndex &lt;- sample(1:length(Markers),1000000)
Markers[misIndex] &lt;- NA

## GSDataQC, not impute ##
QCRes &lt;- GSDataQC(markers = Markers, phenotype = pheData, impute = T)

## GSDataQC, impute with methods mean ##
QCResImpute &lt;- GSDataQC(markers = Markers, phenotype = pheData, impute = T, 
                        imputeMethod = &quot;mean&quot;)
</code></pre>
<h2 id="genomic-selection-and-model-evaluation">Genomic selection and model evaluation</h2>
<h3 id="genotype-to-phenotype-prediction">Genotype-to-phenotype prediction</h3>
<ul>
<li><em><strong><code>G2P</code></strong></em></li>
<li>Description<br/>
The function fits genomic selection models, performs prediction and exports the prediction value of testing sets.</li>
<li>Usage <br/>
<code>G2P(markers,data,fix = NULL,trait,trainIdx,predIdx,modelMethods ="BayesA",
    outputModel = FALSE,NAImpute = FALSE,
    nIter = 1500, burnIn = 500, thin = 5, saveAt = "", S0 = NULL, 
    df0 =5, R2 = 0.5, weights = NULL,verbose = FALSE, rmExistingFiles = TRUE, 
    groups=NULLimportance = FALSE,posPercentage = 0.4,
    BestIndividuals = c("top"),ntree = 500,nodesize = NULL,kernel = c("linear"),
    gamma = 1, cost = 2^(-9), K = 8,eta = 0.7,select = "pls2",fit = "simpls",
    scale.x = FALSE,scale.y = FALSE,eps = 1e-4,maxstep = 100,
    parameters,alpha = 1,X = NULL,family = gaussian(link = identity), lambda = NULL,
    tol.err = 1e-6, tol.conv = 1e-8,epochs = 30, neurons = 4,
    ...)</code></li>
<li>Arguments<ul>
<li><code>markers</code>(numeric, matrix) genootypic data, row represents sample and column represents SNP (feature).Genotypes generally be coded as {0,1,2;0 represents AA(homozygote),2 represents BB(homozygote) and 1 represents AB(heterozygote); missing (NA) alleles are not allowed.</li>
<li><code>data</code>(data.frame) phenotypic data and other information for data set.</li>
<li><code>fix</code>(character, array) names of data that used as fix effects, numeric, NAs are not permitted,</li>
<li><code>testPheno</code>(numeric) the phenotype value of test population individual, default NULL.</li>
<li><code>trait</code>(character) the trait name in data to fit model and perform genotype-to-phenotype prediction.</li>
<li><code>trainIdx</code>(numeric, array) index of training set in marker and data.</li>
<li><code>predIdx</code>(numeric, array)  index of testing/prediction set in marker and data.</li>
<li><code>modelMethods</code>(character) 16 alternative models to fit, including "BayesA", "BayesB", "BayesC", "BL", "BRR", "RKHS", "RRBLUP","LASSO", "SPLS", "SVC", "SVR", "RFR", "RFC", "RR", "RKHS", "BRNN"</li>
<li><code>outputModel</code>(logical) if TRUE, return the list of trained model and prediction results, default FALSE.</li>
<li><code>NAImpute</code>(logical) if true, the missing values in phenotype will be imputed by mean otherwise be removed.</li>
<li><code>nIter,burnIn,thin</code>(integer) the number of iterations, burn-in and thinning,default nIter 7000,burnIn 500,thin 5.</li>
<li><code>saveAt</code>(string) this may include a path and a pre-fix that will be added to the name of the files that are saved as the program runs,default "".  </li>
<li><code>S0,df0</code>(numeric) the scale parameter for the scaled inverse-chi squared prior assigned to the residual variance, only used with Gaussian outcomes.Default S0 NULL,df0 = 5.</li>
<li><code>R2</code>(numeric, (0,1)) the proportion of variance that one expects, a priori, to be explained by the regression. Only used if the hyper-parameters are not specified.Defult 0.5</li>
<li><code>weights</code>(numeric) a vector of weights, may be NULL. If weights is not NULL, the residual variance of each data-point is set to be proportional to the square of the weight. Only used with Gaussian outcomes.</li>
<li><code>verbose</code>(logical) if TRUE the iteration history is printed, default FALSE.</li>
<li><code>rmExistingFiles</code>(logical) if TRUE, removes existing output files from previous runs, default TRUE.</li>
<li><code>groups</code>(factor) a vector of the same length of y that associates observations with groups, each group will have an associated variance component for the error term.</li>
<li><code>ntree</code>(integer) ramdomforest parameter. Number of trees to grow. This should not be set to too small a number, to ensure that every input row gets predicted at least a few times,default 500.</li>
<li><code>nodesize</code>(integer) ramdomforest parameter Minimum size of terminal nodes. Setting this number larger causes smaller trees to be grown (and thus take less time). Note that the default values are different for classification (1) and regression (5).</li>
<li><code>posPercentage</code>(numeric,[0,1]) phenotype of extreme individuals which expected, default 0.4.</li>
<li><code>BestIndividuals</code>(character) the position of expected phenotype in whole phenotypic data set."top","buttom" or "middle",default "top".</li>
<li><code>kernel</code>(numeric) svm parameter the kernel used in training and predicting. You might consider changing some of the following parameters, depending on the kernel type.(linear,polynomial,sigmoid,radial). Default "linear".</li>
<li><code>gamma</code>(numeric)svm parameter parameter needed for all kernels except linear, default 1.</li>
<li><code>cost</code>(numeric)svm parameter cost of constraints violation, default: 2^(-9), it is the 'C'-constant of the regularization term in the Lagrange formulation.</li>
<li><code>K</code>(integer) SPLS parameter: number of hidden components.</li>
<li><code>eta</code>(numeric) SPLS parameter: thresholding parameter. eta should be between 0 and 1.</li>
<li><code>select</code>(character) SPLS parameter: PLS algorithm for variable selection. Alternatives are "pls2" or "simpls". Default is "pls2". </li>
<li><code>fit</code>(character) SPLS parameter: PLS algorithm for model fitting. Alternatives are "kernelpls", "widekernelpls", "simpls", or "oscorespls". Default is "simpls".</li>
<li><code>scale.x</code>(character) SPLS parameter: scale predictors by dividing each predictor variable by its sample standard deviation?</li>
<li><code>scale.y</code>(character) SPLS parameter: scale responses by dividing each response variable by its sample standard deviation?</li>
<li><code>eps</code>(character) SPLS parameter: an effective zero. Default is 1e-4.</li>
<li><code>maxstep</code>(integer) SPLS parameter: maximum number of iterations when fitting direction vectors. Default is 100.</li>
<li><code>alpha</code>(numeric) LASSO model parameter: the elasticnet mixing parameter.Detail in glmnet.</li>
<li><code>family</code>the distribution family of y, see help('family') for more details. </li>
<li><code>lambda</code>the shrinkage parameter determines the amount of shrinkage. Default is NULL meaning that it is to be estimated along with other model parameters.</li>
<li><code>tol.err</code>internal tolerance level for extremely small values; default value is 1e-6.</li>
<li><code>tol.conv</code>tolerance level in convergence; default value is 1e-8.</li>
<li><code>neurons</code>(integer) indicates the number of neurons,defult 4.</li>
<li><code>epochs</code>(integer) maximum number of epochs(iterations) to train, default 30.
  <br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">## load example data ##
data(cubic)

## perform genotype to phenotype prediction with all integrated methods ## 
C2Pres &lt;- G2P(markers = Markers,data = pheData,trait = &quot;DTT&quot;,
               modelMethods = c(&quot;BayesA&quot;, &quot;BayesB&quot;, &quot;BayesC&quot;, &quot;BL&quot;, &quot;BRR&quot;, 
                                &quot;RKHS&quot;, &quot;RRBLUP&quot;,&quot;LASSO&quot;, &quot;SPLS&quot;, &quot;SVC&quot;, &quot;SVR&quot;, &quot;RFR&quot;, &quot;RFC&quot;,
                                &quot;RR&quot;, &quot;RKHS&quot;, &quot;BRNN&quot;),trainIdx = 1:200,predIdx = 201:400)

## perform genotype to phenotype prediction with all integrated methods and output model ## 
C2Pres_outmodel &lt;- G2P(markers = Markers,data = pheData,trait = &quot;DTT&quot;,outputModel = T,
               modelMethods = c(&quot;BayesA&quot;, &quot;BayesB&quot;, &quot;BayesC&quot;, &quot;BL&quot;, &quot;BRR&quot;, 
                                &quot;RKHS&quot;, &quot;RRBLUP&quot;,&quot;LASSO&quot;, &quot;SPLS&quot;, &quot;SVC&quot;, &quot;SVR&quot;, &quot;RFR&quot;, &quot;RFC&quot;,
                                &quot;RR&quot;, &quot;RKHS&quot;, &quot;BRNN&quot;),trainIdx = 1:200,predIdx = 201:400)

## add other information as fix effects ## 
## fix effects not permit NA, so impute NA first.
data &lt;- pheData
data &lt;- apply(data[,-1], 2,function(x){x[which(is.na(x) == T)] &lt;- mean(x,na.rm = T);x)
data &lt;- as.data.frame(data)
data$EW &lt;- pheData$EW

## perform genotype to phenotype prediction ## 
C2Pres_fix &lt;- G2P(markers = Markers,data = data,trait = &quot;EW&quot;,outputModel = T,fix = c(&quot;DTT&quot;,&quot;PH&quot;),
                  modelMethods = c(&quot;BayesA&quot;, &quot;BayesB&quot;, &quot;BayesC&quot;, &quot;BL&quot;, &quot;BRR&quot;, 
                                   &quot;RKHS&quot;, &quot;RRBLUP&quot;,&quot;LASSO&quot;, &quot;SPLS&quot;, &quot;SVC&quot;, &quot;SVR&quot;, &quot;RFR&quot;, &quot;RFC&quot;,
                                   &quot;RR&quot;, &quot;RKHS&quot;, &quot;BRNN&quot;),trainIdx = 1:200,predIdx = 201:400)

</code></pre>
<h3 id="g2p-cross-validation">G2P cross-validation</h3>
<ul>
<li><em><strong><code>G2PCrossValidation</code></strong></em></li>
<li>Description<br/>
The function achieve cross-validation for integrated GS models.</li>
<li>Usage <br/>
<code>G2PCrossValidation(cvSampleList = NULL,cross = 10,times = 1,
                       seed = 1,cpus = 1,markers,data,trait,modelMethods,...)</code></li>
<li>Arguments<ul>
<li><code>cvSampleList</code>(list) the specified cross-validation combination, or a list from function \code{\link{cvSampleIndex.Default NULL and execute a random grouping.</li>
<li><code>cross</code>(integer) the fold of N-fold cross-validation. For example cross = 10 for 10-fold cross-validation.</li>
<li><code>times</code>(integer) the times of repetition. For example, cross = 10 and times = 5 for five times 10-fold cross-validation.</li>
<li><code>seed</code>(integer) random seed for sampling of cross-validation.</li>
<li><code>cpus</code>(integer) number of cpu cores to parallelizing cross-validation(only available in UNIX-like operating systems), default 1.</li>
<li><code>markers</code>(numeric, matrix) row represents sample and column represents SNP (feature).Genotypes generally be coded as {0,1,2;0 represents AA(homozygote),2 represents BB(homozygote) and 1 represents AB(heterozygote); missing (NA) alleles are not allowed.</li>
<li><code>data</code>(data.frame) phenotypic data and other information for data set.</li>
<li><code>trait</code>(character) name of aim trait to perform cross-validation.</li>
<li><code>modelMethods</code>(character) 16 alternative models to fit, including "BayesA, "BayesB, "BayesC", "BL", "BRR", "RKHS", "RRBLUP","LASSO", "SPLS", "SVC", "SVR", "RFR", "RFC", "RR", "RKHS", "BRNN"</li>
<li>Other parameters are same to <em><strong><code>G2P</code></strong></em></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">## load example data ##
data(cubic)

## perform a 2 times 5-fold cross-validation  ##
predlist &lt;- G2PCrossValidation(cross = 5,seed = 1, cpus = 1, times = 2, markers = Markers[1:200,],
                data = pheData[1:200,], trait = &quot;DTT&quot; modelMethods = c(&quot;rrBLUP&quot;, &quot;spls&quot;),
                outputModel = FALSE)
</code></pre>
<h3 id="evaluation">Evaluation</h3>
<ul>
<li><em><strong><code>G2PEvaluation</code></strong></em></li>
<li>Description<br/>
The function consolidates 13 evaluation metrics to evaluete the performance of genomic selection models.</li>
<li>Usage <br/>
<code>G2PEvaluation(realScores, predScores, Probability = FALSE, evalMethod = "RE",
           Beta = 1, BestIndividuals = "top", topAlpha = 1:90,probIndex = NULL)</code></li>
<li>Arguments<ul>
<li><code>realScores</code>(numeric,vector) observation values of validation data.</li>
<li><code>predScores</code>(numeric,vector or matrix) prediction values from genomic selection models.</li>
<li><code>evalMethod</code>(character) alternative evaluation metrics, including "pearson", "kendall", "spearman", "MSE", "R2", "RE", "Kappa", "AUC", "AUCpr", "accuracy", "F1", "meanNDCG", "NDCG".</li>
<li><code>Beta</code>(numeric) the parameter of F-measure or F-score, when beta = 1, meaning F1, Default F1.</li>
<li><code>BestIndividuals</code>(character) the position of expected phenotype in whole phenotypic data set."top","buttom" or "middle",default "top".</li>
<li><code>topAlpha</code>(numeric, array) the proportion of considering excellent samples, default 1:90.</li>
<li><code>globalAlpha</code>(logical) if global metrics (pearson, kendall, spearman, MSE and R2) evaluate by threshold (topAlpha),default FALSE.</li>
<li><code>probIndex</code>(integer) indicate the column index which prediction result is probability. For example, classification model "RFC" locates the fifth column in prediction matrix and then probIndex = 5.
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">data(cubic)
########## prediction ############
C2Pres &lt;- G2P(markers = Markers,data = pheData,trait = &quot;EW&quot;,
              modelMethods = c(&quot;BayesA&quot;, &quot;BRR&quot;, &quot;RRBLUP&quot;,
                               &quot;SPLS&quot;, &quot;SVC&quot;, &quot;RFR&quot;),trainIdx = 1:200,predIdx = 201:400)
########## evaluation ############
evalres &lt;- G2PEvaluation(realScores = C2Pres[,1], predScores = C2Pres[,2:7], 
                           evalMethod = c(&quot;pearson&quot;, &quot;kendall&quot;,&quot;spearman&quot;,&quot;RE&quot;,&quot;Kappa&quot;,
                                          &quot;AUC&quot;,&quot;AUCpr&quot;,&quot;NDCG&quot;,&quot;meanNDCG&quot;,
                                          &quot;MSE&quot;,&quot;R2&quot;,&quot;F1&quot;,&quot;accuracy&quot;),topAlpha = 1:90, probIndex = 5)
</code></pre>
<h2 id="integration-of-prediction-results">Integration of prediction results</h2>
<h3 id="two-results-integration-and-user-defined-integration-more-than-two-models-results">Two results integration and user-defined integration (more than two models results)</h3>
<ul>
<li><em><strong><code>GSMerge</code></strong></em></li>
<li>Description<br/>
The function provides a strategy to integrate the results of two or more methods.</li>
<li>Usage <br/>
<code>GSMerge(predResMat, ratio, autoOptimize = F)</code></li>
<li>Arguments<ul>
<li><code>predResMat</code>(numeric, matrix) the prediction results of algorithms which you want to merge, the first column is the real value of trait.</li>
<li><code>ratio</code>(numeric,array) the weights of every algorithms.</li>
<li><code>autoOptimize</code>(logical) if true, two group results of methods from multiple-results will be selected and then get average (wight 1:1), default FALSE.
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">data(cubic)

## Genotype-to-phenotype ##
G2Pres &lt;- G2P(markers = Markers,data = pheData,trait = &quot;EW&quot;,
              modelMethods = c(&quot;BayesA&quot;, &quot;BRR&quot;, &quot;RRBLUP&quot;,
                               &quot;SPLS&quot;, &quot;RFR&quot;),trainIdx = 1:200,predIdx = 201:400)

## merge 2 model prediction results ## 
GSMergeRes &lt;- GSMerge(predResMat = G2Pres[,c(&quot;realPhenScore&quot;,&quot;RRBLUP&quot;,&quot;SPLS&quot;)], autoOptimize = F,ratio = c(3,7))

## merge 2 model prediction results automatically ## 
GSMergeRes &lt;- GSMerge(predResMat = G2Pres, autoOptimize = T)
</code></pre>
<p>The function provides a strategy to integrate the results of multiple GS methods.</p>
<h3 id="multiple-results-integration">Multiple results integration</h3>
<ul>
<li><em><strong><code>GSEnsemble</code></strong></em></li>
<li>Description<br/>
Results integration of multiple GS methods</li>
<li>Usage <br/>
<code>GSEnsemble(predMat, nrandom = 10, evalMethods, by = 0.1, evaluation = T, topAlpha = 15)</code></li>
<li>Arguments<ul>
<li><code>predMat</code>(numeric, matrix) the prediction results of algorithms which you want to merge, the first column is the real value of trait.</li>
<li><code>nrandom</code>(integer) the repeat number of stacking, default 10.</li>
<li><code>evalMethods</code>(character) ensemble base which evaluation methods.</li>
<li><code>by</code>(numeric,(0,1)) the radio window of ensemble, the smaller "by", the higher accuracy of ensemble. Default 0.1.</li>
<li><code>evaluation</code>(logical) if evaluate finalMat with evalMethods, default TRUE.</li>
<li><code>topAlpha</code>(numeric) the parameter of threshold evaluation methods, see also function evaluateGS. In this function, indicates the best ensemble base threshold when evalMethods is threshold methods.
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">## load example data ##
data(cubic)

## Genotype-to-phenotype ##
G2Pres &lt;- G2P(markers = Markers,data = pheData,trait = &quot;EW&quot;,
              modelMethods = c(&quot;BayesA&quot;, &quot;BRR&quot;, &quot;RRBLUP&quot;,
                               &quot;SPLS&quot;, &quot;RFR&quot;),trainIdx = 1:200,predIdx = 201:400)

## ensemble more than 2 prediction results of models automatically ## 
GSEnsembleRes &lt;- GSEnsemble(predMat = G2Pres, nrandom = 10, evalMethods = &quot;pearson&quot;,
                          by = 0.1, evaluation = T)
## new prediction of genotype-to-phenotype ##
G2PresNew &lt;- G2P(markers = Markers,data = pheData,trait = &quot;EW&quot;,
              modelMethods = c(&quot;BayesA&quot;, &quot;BRR&quot;, &quot;RRBLUP&quot;,
                               &quot;SPLS&quot;, &quot;RFR&quot;),trainIdx = 1:200,predIdx = 401:600)

## apply GSEnsemble rotio to new prediction ##                       
GSEnsembleModelRes &lt;- GSMerge(predResMat = G2PresNew,ratio = GSEnsembleRes$BestWeight, autoOptimize = F)
</code></pre>
<h2 id="refinement-design-of-training-dataset">Refinement design of training dataset</h2>
<ul>
<li><em><strong><code>TSRefine</code></strong></em></li>
<li>Description<br/>
Refining the training set for selective phenotyping to reduce cost.</li>
<li>Usage <br/>
<code>TSRefine(markers,candidates,test = NULL,ntosel,method = "PEVmean",
         npop = 100, nelite =5 ,mutprob = .8,niterations = 500,
         cores = 1,lambda = NULL,sequent = FALSE,visulization = F)</code></li>
<li>Arguments<ul>
<li><code>markers</code>(numeric, matrix) row represents sample and column represents SNP (feature).Genotypes generally be coded as {0,1,2;0 represents AA(homozygote),2 represents BB(homozygote) and 1 represents AB(heterozygote); missing (NA) alleles are not allowed.</li>
<li><code>candidates</code>(charcter array) names of condidates population.</li>
<li><code>test</code>(character, array) names of testing or predicted set, defult NULL, if NULL, execute reference-free refinement, else execute reference-based refinement.</li>
<li><code>ntosel</code>(integer) number of samples to select in refinement.</li>
<li><code>method</code>(character) alternativeal algorithms for training set refinement or selection, including "PEVmean", "CDmean", "sim", or "RD". Default is "PEVmean".</li>
<li><code>npop</code>(integer) genetic algorithm parameter, number of solutions at each iteration.</li>
<li><code>nelite</code>(integer) genetic algorithm parameter, number of solutions selected as elite parents which will generate the next set of solutions.</li>
<li><code>mutprob</code>(numeric) genetic algorithm parameter, probability of mutation for each generated soluation.</li>
<li><code>niterations</code>(integer) genetic algorithm parameter, number of iterations.</li>
<li><code>cores</code>(integer) number of CPU cores to use.</li>
<li><code>lambda</code>(numeric) scalar shrinkage parameter (λ&gt;0).</li>
<li><code>sequent</code>(logical) phenotype selection parameter, if TRUE, perform sequential optimization, default FALSE.
<br/></li>
</ul>
</li>
</ul>
<p>Example</p>
<pre><code class="language-R">## load example data ##
data(cubic)

## reference-free refinement ##
G &lt;- Markers[1:300,]
TSR_SP &lt;- TSRefine(markers = G, candidates = rownames(G),
              test = NULL, ntosel = 100,cores = 1,visulization = T)

## reference-based refinement (PEV-mean) ##
TSR_PEVmean &lt;- TSRefine(markers = G, method = &quot;PEVmean&quot;, candidates = rownames(G)[1:200],
                       test = rownames(Markers)[201:300], ntosel = 50,cores = 1,visulization = T)
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../QuickStart/" class="btn btn-neutral float-left" title="QuickStart"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../about/" class="btn btn-neutral float-right" title="About">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/GOVS-env/G2P/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../QuickStart/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../about/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
